File: backend/app/api/routes/submissions.py
from datetime import datetime, timedelta, timezone
from typing import Any, Literal
from uuid import UUID

import asyncpg
from fastapi import APIRouter, Depends, HTTPException, Request, status
from fastapi.responses import Response
from pydantic import BaseModel, Field
from slowapi import Limiter
from slowapi.util import get_remote_address

from app.database.repositories import bounties as bounties_repo
from app.database.repositories import submissions as submissions_repo
from app.dependencies import (
    get_current_user,
    get_pool,
    require_org_admin_or_moderator,
    require_participant,
)
from app.models.user import User
from app.services import storage_service as storage_svc
from app.services import submission_service
from app.services.storage_service import StorageService, get_storage_service

limiter = Limiter(key_func=get_remote_address)

# ── Participant router ────────────────────────────────────────────────────────
# Prefix: /bounties/{bounty_id}/submissions
participant_router = APIRouter(tags=["submissions"])

# ── Org router ────────────────────────────────────────────────────────────────
# Prefix: /orgs/{org_id}
org_router = APIRouter(tags=["org-submissions"])


# ── Request / Response models ─────────────────────────────────────────────────


class UploadUrlResponse(BaseModel):
    signed_url: str
    upload_token: UUID
    upload_url_expires_at: datetime


class SubmissionCreateRequest(BaseModel):
    submission_type: Literal["zip", "github_url", "drive_url"]
    upload_token: UUID | None = None
    external_url: str | None = None
    description: str = Field(min_length=10, max_length=5000)


class SubmissionUpdateRequest(BaseModel):
    description: str | None = Field(default=None, min_length=10, max_length=5000)
    external_url: str | None = None
    upload_token: UUID | None = None          # new zip file for replacement
    new_submission_type: Literal["zip", "github_url", "drive_url"] | None = None


class StatusUpdateRequest(BaseModel):
    status: Literal["under_review", "rejected", "scored"]
    review_notes: str | None = None


# ── Participant endpoints ─────────────────────────────────────────────────────

# IMPORTANT: /upload-url and /mine are registered BEFORE /{sub_id} to avoid
# FastAPI matching literal segments as UUIDs.


@participant_router.post(
    "/bounties/{bounty_id}/submissions/upload-url",
    response_model=UploadUrlResponse,
    status_code=status.HTTP_201_CREATED,
)
@limiter.limit("5/minute")
async def get_upload_url(
    request: Request,
    bounty_id: UUID,
    user: User = Depends(require_participant),
    pool: asyncpg.Pool = Depends(get_pool),
    storage: StorageService = Depends(get_storage_service),
) -> Any:
    bounty = await bounties_repo.get_by_id(pool, bounty_id)
    if not bounty or bounty["status"] != "open":
        raise HTTPException(409, "Bounty is not open")
    if "zip" not in bounty["submission_formats"]:
        raise HTTPException(400, "This bounty does not accept zip submissions")
    await submission_service.check_submission_limit(pool, bounty_id, user.id, bounty)

    expires_at = datetime.now(timezone.utc) + timedelta(minutes=10)
    path, sub_id = submission_service.generate_upload_path(bounty_id, user.id)
    await submissions_repo.create_pending(pool, sub_id, bounty_id, user.id, path, expires_at)

    signed = await storage.create_signed_upload_url(path)
    return UploadUrlResponse(
        signed_url=signed.signed_url,
        upload_token=sub_id,
        upload_url_expires_at=expires_at,
    )


@participant_router.get("/bounties/{bounty_id}/submissions/mine/download-url")
async def get_my_submission_download_url(
    bounty_id: UUID,
    user: User = Depends(require_participant),
    pool: asyncpg.Pool = Depends(get_pool),
    storage: StorageService = Depends(get_storage_service),
) -> dict:
    sub = await submissions_repo.get_mine(pool, bounty_id, user.id)
    if not sub:
        raise HTTPException(404, "No submission found")
    if not sub.get("file_path"):
        raise HTTPException(400, "Submission has no file")
    signed_url = await storage.create_signed_download_url(sub["file_path"])
    return {"download_url": signed_url}


@participant_router.get("/bounties/{bounty_id}/submissions/mine")
async def get_my_submission(
    bounty_id: UUID,
    user: User = Depends(require_participant),
    pool: asyncpg.Pool = Depends(get_pool),
) -> Any:
    sub = await submissions_repo.get_mine(pool, bounty_id, user.id)
    if not sub:
        raise HTTPException(404, "No submission found")
    return sub


@participant_router.post(
    "/bounties/{bounty_id}/submissions",
    status_code=status.HTTP_201_CREATED,
)
async def create_submission(
    bounty_id: UUID,
    payload: SubmissionCreateRequest,
    user: User = Depends(require_participant),
    pool: asyncpg.Pool = Depends(get_pool),
) -> Any:
    bounty = await bounties_repo.get_by_id(pool, bounty_id)
    if not bounty or bounty["status"] != "open":
        raise HTTPException(409, "Bounty is not open")
    if payload.submission_type not in bounty["submission_formats"]:
        raise HTTPException(400, f"Bounty does not accept {payload.submission_type} submissions")

    if payload.submission_type in ("github_url", "drive_url"):
        submission_service.validate_submission_url(payload.submission_type, payload.external_url)

    file_path = None
    if payload.submission_type == "zip":
        file_path = await submission_service.resolve_upload_token(
            pool, payload.upload_token, user.id, bounty_id
        )

    try:
        return await submissions_repo.create(
            pool,
            bounty_id=bounty_id,
            user_id=user.id,
            submission_type=payload.submission_type,
            file_path=file_path,
            external_url=payload.external_url,
            description=payload.description,
            upload_token=payload.upload_token,
            max_per_user=bounty["max_submissions_per_user"],
        )
    except asyncpg.UniqueViolationError:
        raise HTTPException(409, "You already have an active submission for this bounty")


@participant_router.post(
    "/bounties/{bounty_id}/submissions/{sub_id}/replace-url",
    response_model=UploadUrlResponse,
    status_code=status.HTTP_201_CREATED,
)
async def get_replace_url(
    bounty_id: UUID,
    sub_id: UUID,
    user: User = Depends(require_participant),
    pool: asyncpg.Pool = Depends(get_pool),
    storage: StorageService = Depends(get_storage_service),
) -> Any:
    """Get a signed upload URL for replacing or switching to a zip submission."""
    sub = await submissions_repo.get_by_id_and_bounty(pool, sub_id, bounty_id)
    if not sub or sub["user_id"] != user.id:
        raise HTTPException(404, "Submission not found")
    if sub["status"] != "pending":
        raise HTTPException(409, "Can only replace file for pending submissions")
    bounty = await bounties_repo.get_by_id(pool, bounty_id)
    if not bounty or "zip" not in bounty["submission_formats"]:
        raise HTTPException(400, "This bounty does not accept zip submissions")

    expires_at = datetime.now(timezone.utc) + timedelta(minutes=10)
    path, token_sub_id = submission_service.generate_upload_path(bounty_id, user.id)
    await submissions_repo.create_pending(pool, token_sub_id, bounty_id, user.id, path, expires_at)
    signed = await storage.create_signed_upload_url(path)
    return UploadUrlResponse(
        signed_url=signed.signed_url,
        upload_token=token_sub_id,
        upload_url_expires_at=expires_at,
    )


@participant_router.patch("/bounties/{bounty_id}/submissions/{sub_id}")
async def update_submission(
    bounty_id: UUID,
    sub_id: UUID,
    payload: SubmissionUpdateRequest,
    user: User = Depends(require_participant),
    pool: asyncpg.Pool = Depends(get_pool),
    storage: StorageService = Depends(get_storage_service),
) -> Any:
    sub = await submissions_repo.get_by_id_and_bounty(pool, sub_id, bounty_id)
    if not sub:
        raise HTTPException(404, "Submission not found")
    if sub["user_id"] != user.id:
        raise HTTPException(403, "Not your submission")
    if sub["status"] != "pending":
        raise HTTPException(409, "Cannot edit a submission that is under review or scored")

    target_type = payload.new_submission_type or sub["submission_type"]

    # If switching type, verify bounty supports the target
    if target_type != sub["submission_type"]:
        bounty = await bounties_repo.get_by_id(pool, bounty_id)
        if not bounty or target_type not in bounty["submission_formats"]:
            raise HTTPException(400, f"Bounty does not accept {target_type} submissions")

    fields: dict = {}
    if payload.description is not None:
        fields["description"] = payload.description

    if target_type == "zip":
        if payload.upload_token:
            new_path = await submission_service.resolve_upload_token(
                pool, payload.upload_token, user.id, bounty_id
            )
            old_path = sub.get("file_path")
            fields["file_path"] = new_path
            if target_type != sub["submission_type"]:
                fields["external_url"] = None  # clear old URL
            # clean up old file
            if old_path:
                try:
                    await storage.delete_file(old_path)
                except Exception:
                    pass
            await submissions_repo.delete_pending(pool, payload.upload_token)
        elif target_type != sub["submission_type"]:
            raise HTTPException(400, "upload_token required when switching to zip submission type")
        # else: zip→zip with no new file — only description changes
    else:
        # URL type
        if payload.external_url is not None:
            submission_service.validate_submission_url(target_type, payload.external_url)
            fields["external_url"] = payload.external_url
        elif target_type != sub["submission_type"]:
            raise HTTPException(400, "external_url required when switching to URL submission type")
        # switching away from zip: clear file_path and delete old file
        if sub["submission_type"] == "zip":
            fields["file_path"] = None
            old_path = sub.get("file_path")
            if old_path:
                try:
                    await storage.delete_file(old_path)
                except Exception:
                    pass

    if target_type != sub["submission_type"]:
        fields["submission_type"] = target_type

    return await submissions_repo.update_fields(pool, sub_id, fields)


# ── Org endpoints ─────────────────────────────────────────────────────────────


@org_router.get("/orgs/{org_id}/bounties/{bounty_id}/submissions")
async def list_submissions(
    org_id: UUID,
    bounty_id: UUID,
    status_filter: str | None = None,
    user: User = Depends(require_org_admin_or_moderator),
    pool: asyncpg.Pool = Depends(get_pool),
) -> list:
    # Verify bounty belongs to org
    bounty = await bounties_repo.get_for_org(pool, bounty_id, org_id)
    if not bounty:
        raise HTTPException(404, "Bounty not found")
    return await submissions_repo.list_for_bounty(pool, bounty_id, org_id, status_filter)


@org_router.patch("/orgs/{org_id}/submissions/{sub_id}/status")
async def update_submission_status(
    org_id: UUID,
    sub_id: UUID,
    payload: StatusUpdateRequest,
    user: User = Depends(require_org_admin_or_moderator),
    pool: asyncpg.Pool = Depends(get_pool),
) -> Any:
    sub = await submissions_repo.get_for_org(pool, sub_id, org_id)
    if not sub:
        raise HTTPException(404, "Submission not found")
    submission_service.validate_status_transition(sub["status"], payload.status)
    return await submissions_repo.update_status(pool, sub_id, payload.status, payload.review_notes)


@org_router.get("/orgs/{org_id}/submissions/{sub_id}/download-url")
async def get_download_url(
    org_id: UUID,
    sub_id: UUID,
    user: User = Depends(require_org_admin_or_moderator),
    pool: asyncpg.Pool = Depends(get_pool),
    storage: StorageService = Depends(get_storage_service),
) -> dict:
    sub = await submissions_repo.get_for_org(pool, sub_id, org_id)
    if not sub:
        raise HTTPException(404, "Submission not found")
    if not sub.get("file_path"):
        raise HTTPException(400, "Submission has no file")
    signed_url = await storage.create_signed_download_url(sub["file_path"])
    return {"download_url": signed_url}

File: ui/src/app/actions/mutations/submissions.ts
'use server'
import { requireAuth, requireOrgMember } from '../_auth'

const GITHUB_URL_RE =
  /^https:\/\/github\.com\/[A-Za-z0-9]([A-Za-z0-9_-]{0,38}[A-Za-z0-9])?\/[A-Za-z0-9][A-Za-z0-9._-]{0,98}[A-Za-z0-9]$/

const DRIVE_URL_RE =
  /^https:\/\/drive\.google\.com\/(file\/d\/[A-Za-z0-9_-]+(\/[^?#]*)?|drive\/folders\/[A-Za-z0-9_-]+(\/[^?#]*)?|open\?id=[A-Za-z0-9_-]+)$/

const STORAGE_PATH_RE =
  /^submissions\/[0-9a-f]{8}-[0-9a-f]{4}-4[0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}\/[0-9a-f]{8}-[0-9a-f]{4}-4[0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}\/[0-9a-f]{8}-[0-9a-f]{4}-4[0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}\.zip$/i

const VALID_TRANSITIONS: Record<string, string[]> = {
  pending: ['under_review', 'rejected'],
  under_review: ['rejected', 'scored'],
}

export async function createSubmission(data: {
  bountyId: string
  uploadToken?: string
  externalUrl?: string
  submissionType: 'zip' | 'github_url' | 'drive_url'
  description?: string
}) {
  const auth = await requireAuth()
  if (!auth.ok) return { error: auth.error }

  const { bountyId, uploadToken, externalUrl, submissionType, description } = data

  // Validate URL submissions
  if (submissionType === 'github_url') {
    if (!externalUrl || !GITHUB_URL_RE.test(externalUrl)) {
      return { error: 'Invalid GitHub repository URL' }
    }
  } else if (submissionType === 'drive_url') {
    if (!externalUrl || !DRIVE_URL_RE.test(externalUrl)) {
      return { error: 'Invalid Google Drive URL' }
    }
  }

  // Fetch bounty details
  const { data: bounty } = await auth.admin
    .from('bounties')
    .select('status, submission_formats, max_submissions_per_user')
    .eq('id', bountyId)
    .single()

  if (!bounty) return { error: 'Bounty not found' }
  if (bounty.status !== 'open') return { error: 'This bounty is not open for submissions' }
  if (!bounty.submission_formats.includes(submissionType)) {
    return { error: `Submission type ${submissionType} is not accepted for this bounty` }
  }

  // max_submissions_per_user enforcement (non-rejected submissions)
  if (bounty.max_submissions_per_user) {
    const { count } = await auth.admin
      .from('submissions')
      .select('id', { count: 'exact', head: true })
      .eq('bounty_id', bountyId)
      .eq('user_id', auth.user.id)
      .not('status', 'in', '(rejected,upload_pending)')

    if ((count ?? 0) >= bounty.max_submissions_per_user) {
      return { error: `Maximum ${bounty.max_submissions_per_user} submissions allowed per user` }
    }
  }

  if (submissionType === 'zip') {
    // Validate upload token
    if (!uploadToken) return { error: 'Upload token is required for zip submissions' }

    const { data: pending } = await auth.admin
      .from('submissions')
      .select('id, file_path, upload_token_expires_at')
      .eq('id', uploadToken)
      .eq('user_id', auth.user.id)
      .eq('bounty_id', bountyId)
      .eq('status', 'upload_pending')
      .single()

    if (!pending) return { error: 'Invalid or expired upload token' }
    if (
      pending.upload_token_expires_at &&
      new Date(pending.upload_token_expires_at) < new Date()
    ) {
      return { error: 'Upload token has expired' }
    }
    if (!pending.file_path || !STORAGE_PATH_RE.test(pending.file_path)) {
      return { error: 'Invalid storage path' }
    }

    // Promote upload_pending to pending
    const { error } = await auth.admin
      .from('submissions')
      .update({
        status: 'pending',
        submission_type: 'zip',
        description: description ?? '',
        submitted_at: new Date().toISOString(),
        updated_at: new Date().toISOString(),
      })
      .eq('id', uploadToken)

    if (error) {
      if (error.code === '23505') {
        return { error: 'You already have an active submission for this bounty' }
      }
      return { error: error.message }
    }

    return { data: { id: uploadToken } }
  }

  // URL-based submission
  const { data: submission, error } = await auth.admin
    .from('submissions')
    .insert({
      bounty_id: bountyId,
      user_id: auth.user.id,
      status: 'pending',
      submission_type: submissionType,
      external_url: externalUrl,
      description: description ?? '',
      submitted_at: new Date().toISOString(),
    })
    .select('id')
    .single()

  if (error) {
    if (error.code === '23505') {
      return { error: 'You already have an active submission for this bounty' }
    }
    return { error: error.message }
  }

  return { data: submission }
}

export async function updateSubmission(
  submissionId: string,
  updates: {
    description?: string
    submissionType?: 'zip' | 'github_url' | 'drive_url'
    externalUrl?: string
    uploadToken?: string
  }
) {
  const auth = await requireAuth()
  if (!auth.ok) return { error: auth.error }

  const { data: sub } = await auth.admin
    .from('submissions')
    .select('id, status, submission_type, file_path, bounty_id')
    .eq('id', submissionId)
    .eq('user_id', auth.user.id)
    .single()

  if (!sub) return { error: 'Submission not found' }
  if (sub.status !== 'pending') return { error: 'Only pending submissions can be edited' }

  if (updates.submissionType === 'github_url') {
    if (!updates.externalUrl || !GITHUB_URL_RE.test(updates.externalUrl)) {
      return { error: 'Invalid GitHub repository URL' }
    }
  } else if (updates.submissionType === 'drive_url') {
    if (!updates.externalUrl || !DRIVE_URL_RE.test(updates.externalUrl)) {
      return { error: 'Invalid Google Drive URL' }
    }
  }

  // Switching FROM zip: delete old storage file
  if (sub.submission_type === 'zip' && updates.submissionType && updates.submissionType !== 'zip') {
    if (sub.file_path) {
      await auth.admin.storage.from('submission-zips').remove([sub.file_path])
    }
  }

  const { error } = await auth.admin
    .from('submissions')
    .update({
      ...(updates.description !== undefined && { description: updates.description }),
      ...(updates.submissionType && { submission_type: updates.submissionType }),
      ...(updates.externalUrl !== undefined && { external_url: updates.externalUrl }),
      updated_at: new Date().toISOString(),
    })
    .eq('id', submissionId)

  if (error) return { error: error.message }
  return { data: { success: true } }
}

export async function updateSubmissionStatus(
  submissionId: string,
  orgId: string,
  newStatus: 'under_review' | 'rejected' | 'scored',
  reviewNotes?: string
) {
  const auth = await requireOrgMember(orgId)
  if (!auth.ok) return { error: auth.error }

  const { data: sub } = await auth.admin
    .from('submissions')
    .select('status')
    .eq('id', submissionId)
    .single()

  if (!sub) return { error: 'Submission not found' }

  const allowed = VALID_TRANSITIONS[sub.status] ?? []
  if (!allowed.includes(newStatus)) {
    return { error: `Cannot transition from ${sub.status} to ${newStatus}` }
  }

  const { error } = await auth.admin
    .from('submissions')
    .update({
      status: newStatus,
      ...(reviewNotes !== undefined && { review_notes: reviewNotes }),
      updated_at: new Date().toISOString(),
    })
    .eq('id', submissionId)

  if (error) return { error: error.message }
  return { data: { success: true } }
}

export async function getDownloadUrl(submissionId: string) {
  const auth = await requireAuth()
  if (!auth.ok) return { error: auth.error }

  const { data: sub } = await auth.admin
    .from('submissions')
    .select('file_path, user_id, bounty_id')
    .eq('id', submissionId)
    .single()

  if (!sub || !sub.file_path) return { error: 'Submission not found or not a zip submission' }

  // Check caller is submitter or org admin/moderator
  const isOwner = sub.user_id === auth.user.id
  if (!isOwner) {
    const { data: membership } = await auth.admin
      .from('bounties')
      .select('org_members!inner(role)')
      .eq('id', sub.bounty_id)
      .eq('org_members.user_id', auth.user.id)
      .single()

    if (!membership) return { error: 'Forbidden' }
  }

  const { data: signed, error } = await auth.admin.storage
    .from('submission-zips')
    .createSignedUrl(sub.file_path, 3600) // 1-hour expiry

  if (error || !signed) return { error: 'Failed to generate download URL' }
  return { data: { url: signed.signedUrl } }
}
